core/drivers/crypto/caam/caam_key.c

```diff
@@ -0,0 +1,746 @@
+		return status;
+	}
+
+	memcpy(blob.buf.data, data + get_key_buf_offset(data, size),
+	       get_key_buf_size(data, size));
+
+	/* Set destination key */
+	key->key_type = blob.key_type;
+	key->sec_size = blob.sec_size;
+	key->is_blob = false;
+
+	/* De-blob operation */
+	status = caam_key_operation_blob(&blob, key);
+	if (status) {
+		KEY_TRACE("De-blob operation fail");
+		goto out;
+	}
+
+	KEY_TRACE("Deserialization binary buffer done");
+	caam_key_dump("Deserialization output key", key);
+out:
+	caam_key_free(&blob);
+	return status;
+}
+
+enum caam_status caam_key_serialize_to_bin(uint8_t *data, size_t size,
+					   const struct caamkey *key)
+{
+	struct caam_key_serialized key_ser = { };
+	struct caamkey blob = { };
+	enum caam_status status = CAAM_FAILURE;
+	size_t serialized_size = 0;
+
+	assert(data && size && key);
+
+	caam_key_dump("Serialization input key", key);
+
+	/* If the key is plain text, just copy key to buffer */
+	if (key->key_type == CAAM_KEY_PLAIN_TEXT) {
+		if (size < key->buf.length) {
+			KEY_TRACE("Buffer is too short");
+			return CAAM_SHORT_BUFFER;
+		}
+
+		memcpy(data, key->buf.data, key->buf.length);
+
+		return CAAM_NO_ERROR;
+	}
+
+	/* The input key must not be a blob */
+	assert(!key->is_blob);
+
+	/* Blob the given key for serialization and export */
+	blob.is_blob = true;
+	blob.sec_size = key->sec_size;
+	blob.key_type = key->key_type;
+
+	/*
+	 * Check if the destination is big enough for the black blob buffer and
+	 * header.
+	 */
+	status = caam_key_serialized_size(&blob, &serialized_size);
+	if (status)
+		goto out;
+
+	if (size < serialized_size) {
+		KEY_TRACE("Destination buffer is too short %zu < %zu", size,
+			  serialized_size);
+		status = CAAM_OUT_MEMORY;
+		goto out;
+	}
+
+	/* Blob the given key */
+	status = caam_key_operation_blob(key, &blob);
+	if (status) {
+		KEY_TRACE("Blob operation fail");
+		goto out;
+	}
+
+	/* Copy the header to destination */
+	key_ser.magic_number = MAGIC_NUMBER;
+	key_ser.key_type = blob.key_type;
+	key_ser.sec_size = blob.sec_size;
+	memcpy(data, &key_ser, sizeof(key_ser));
+
+	/* Copy the key buffer */
+	memcpy(data + sizeof(key_ser), blob.buf.data, blob.buf.length);
+
+	KEY_DUMPBUF("Key data", data, size);
+out:
+	caam_key_free(&blob);
+
+	return status;
+}
+
+enum caam_status caam_key_serialized_size(const struct caamkey *key,
+					  size_t *size)
+{
+	assert(key && size);
+
+	/* For a plain text key, the serialized key is identical to the key */
+	*size = key->buf.length;
+
+	/*
+	 * For black keys, the serialized key includes the header and must be
+	 * in a blob format
+	 */
+	if (key->key_type != CAAM_KEY_PLAIN_TEXT) {
+		size_t alloc = 0;
+		struct caam_key_serialized key_serialized = { };
+		const struct caamkey tmp = {
+			.key_type = key->key_type,
+			.sec_size = key->sec_size,
+			.is_blob = true,
+		};
+
+		alloc = caam_key_get_alloc_size(&tmp);
+		if (!alloc)
+			return CAAM_FAILURE;
+
+		*size = alloc + sizeof(key_serialized);
```

Comment:
`sizeof(struct caam_key_serialized)`, and remove the otherwise unused `key_serialized`.

Reply:
Indeed